### PR TITLE
Quic update identity

### PR DESCRIPTION
**Pull request by ryleung-solana**
#### Problem
See #31557 

#### Summary of Changes
Implements key hotswapping for the quic server and connection cache

Fixes #

### CODE REVIEW

File: connection-cache/src/connection_cache.rs [hunk 1]
```diff
@@ -137,6 +138,12 @@ where
             .unwrap()
     }
 
+    pub fn update_key(&self, key: &Keypair) {
+        let mut map = self.map.write().unwrap();
+        map.clear();
+        let mut connection_manager = self.connection_manager.write().unwrap();
+        connection_manager.update_key(key);
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply connection-cache/src/connection_cache.rs:145]: As far as I can see, the order is always self.map -> self.connection_manager so this should not cause a deadlock.

File: quic-client/src/lib.rs [hunk 2]
```diff
@@ -140,9 +142,15 @@ impl QuicConfig {
...
     ) -> Result<(), RcgenError> {
-        let (cert, priv_key) = new_self_signed_tls_certificate(keypair, ipaddr)?;
+        let (cert, priv_key) = if let Some(ip) = ipaddr {
+            let res = new_self_signed_tls_certificate(keypair, ip)?;
+            self.ip = ip;
+            res
+        } else {
+            new_self_signed_tls_certificate(keypair, self.ip)?
+        };
         self.client_certificate = Arc::new(QuicClientCertificate {
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply quic-client/src/lib.rs:154]: See QuicLazyInitializedEndpoint::create_endpoint

File: quic-client/src/lib.rs [hunk 3]
```diff
@@ -212,6 +220,10 @@ impl ConnectionManager for QuicConnectionManager {
     fn new_connection_config(&self) -> QuicConfig {
         self.connection_config.clone()
     }
+
+    fn update_key(&mut self, key: &Keypair) {
+        let _ = self.connection_config.update_client_certificate(key, None);
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply quic-client/src/lib.rs]: Good catch - logged

File: connection-cache/src/connection_cache.rs [hunk 4]
```diff
@@ -47,7 +48,7 @@ pub struct ConnectionCache<
 > {
     name: &'static str,
     map: Arc<RwLock<IndexMap<SocketAddr, /*ConnectionPool:*/ R>>>,
-    connection_manager: Arc<S>,
+    connection_manager: Arc<RwLock<S>>,
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply connection-cache/src/connection_cache.rs:51]: Agreed that this is not ideal. Since the map is always locked when connection_manager is locked (but not vice versa), it should be possible to just use 1 lock to protect both. However, while I have tried to remove this lock, I have not been able to make Rust happy (non-buiilding code at https://github.com/ryleung-solana/solana/tree/quic-update-identity-test if you have suggestions).

File: quic-client/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (behzadnouri) [reply quic-client/src/lib.rs:155]: Right, but that is only called once:
https://github.com/solana-labs/solana/blob/a96be5d2f/quic-client/src/nonblocking/quic_client.rs#L142-L143
and the cached in:
```rust
endpoint: OnceCell<Arc<Endpoint>>,
```
https://github.com/solana-labs/solana/blob/a96be5d2f/quic-client/src/nonblocking/quic_client.rs#L70

File: quic-client/src/lib.rs [hunk 5]
```diff
@@ -140,9 +142,15 @@ impl QuicConfig {
...
+        ipaddr: Option<IpAddr>,
     ) -> Result<(), RcgenError> {
-        let (cert, priv_key) = new_self_signed_tls_certificate(keypair, ipaddr)?;
+        let (cert, priv_key) = if let Some(ip) = ipaddr {
+            let res = new_self_signed_tls_certificate(keypair, ip)?;
+            self.ip = ip;
+            res
+        } else {
+            new_self_signed_tls_certificate(keypair, self.ip)?
+        };
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply quic-client/src/lib.rs:153]: I found it more convenient to just store the IP in QuicConfig since that's where we the socket anyway, and bubbling it up just to store it somewhere else seems like a little too much added complexity.

File: quic-client/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (behzadnouri) [reply quic-client/src/lib.rs:154]: @ryleung-solana you marked this resolved, but I do not know what have you changed to resolve this.

File: streamer/src/quic.rs [hunk 6]
```diff
@@ -404,7 +416,7 @@ pub fn spawn_server(
     max_unstaked_connections: usize,
     wait_for_chunk_timeout: Duration,
     coalesce: Duration,
-) -> Result<(Endpoint, thread::JoinHandle<()>), QuicServerError> {
+) -> Result<(Endpoint, thread::JoinHandle<()>, Arc<EndpointKeyUpdater>), QuicServerError> {
```
[the remark this answers is not in the document]
Comment (behzadnouri) [reply streamer/src/quic.rs]: Same here, marked resolved, but looks unchanged.

File: connection-cache/src/connection_cache.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (behzadnouri) [reply connection-cache/src/connection_cache.rs]: Can we change the order of locking and release the lock so we don't hold the lock on both?
something like:
```rust
connection_manager.write().unwrap().update_key(key);
self.map.write().unwrap().clear();
```
This might discard some ok connections created between the 1st and the 2nd lock but I think that is better than the possibility of a dead-lock in the future.

File: quic-client/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply quic-client/src/lib.rs:154]: In QuicConfig::create_endpoint we create a new QuicLazyInitializedEndpoint struct. This, combined with clearing the connection cache means that the first time we get a new connection from the connection cache, we will call set_default_client_config on the Endpoint

File: connection-cache/src/connection_cache.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply connection-cache/src/connection_cache.rs]: I just added a comment stating the required locking order. I think that is the better solution, since it ensures stricter correctness, and it's quite common to simply establish a locking order to prevent deadlocks, rather than not lock multiple locks at all.

File: streamer/src/quic.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply streamer/src/quic.rs]: Looks changed to me...

[the remark this answers is not in the document]
Comment (ryleung-solana) [reply streamer/src/quic.rs]: Looks changed to me...

File: connection-cache/src/connection_cache.rs [hunk 7]
```diff
@@ -38,6 +38,9 @@ pub trait ConnectionManager: Send + Sync + 'static {
 
     fn new_connection_pool(&self) -> Self::ConnectionPool;
     fn new_connection_config(&self) -> Self::NewConnectionConfig;
+    fn update_key(&mut self, _key: &Keypair) -> Result<(), Box<dyn std::error::Error>> {
+        Ok(())
+    }
```
Comment (behzadnouri) [inline connection-cache/src/connection_cache.rs:43]: This looks like a bad default implementation, and I am not sure we need a default implementation here.
Can you please remove it if not needed?

File: connection-cache/src/connection_cache.rs [hunk 8]
```diff
@@ -96,8 +102,8 @@ where
         Self {
             name,
             map,
-            stats,
             connection_manager,
+            stats,
```
Comment (behzadnouri) [inline connection-cache/src/connection_cache.rs:106]: revert this change?

File: connection-cache/src/connection_cache.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (behzadnouri) [reply connection-cache/src/connection_cache.rs]: There is no way to enforce the comment.
Someone might change/add code later unaware of that comment there.
What is the problem with my suggestion above? https://github.com/solana-labs/solana/pull/33865#discussion_r1390311487

File: validator/src/admin_rpc_service.rs [hunk 9]
```diff
@@ -64,6 +66,14 @@ impl AdminRpcRequestMetadata {
             ))
         }
     }
+
+    pub fn update_keypair(&self, key: &Keypair) {
+        for n in self.notifies.iter() {
+            if let Err(err) = n.update_key(key) {
+                warn!("Error updating network layer keypair: {}", err);
```
Comment (behzadnouri) [inline validator/src/admin_rpc_service.rs:73]: This should be `error!`.
also, https://rust-lang.github.io/rust-clippy/master/#/uninlined_format_args

File: validator/src/admin_rpc_service.rs [hunk 10]
```diff
@@ -64,6 +66,14 @@ impl AdminRpcRequestMetadata {
             ))
         }
     }
+
+    pub fn update_keypair(&self, key: &Keypair) {
```
Comment (behzadnouri) [inline validator/src/admin_rpc_service.rs:70]: Does this need to be `pub`?

File: connection-cache/src/connection_cache.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply connection-cache/src/connection_cache.rs]: > This might discard some ok connections created between the 1st and the 2nd lock

There's also no way to enforce that only 1 of the locks can be held at a time. It's not uncommon to just specify a locking order that future programmers must follow when dealing with multiple locks. Not a massive deal, but since we have 2 locks anyway, I would prefer to do the most strictly correct thing and clear the cache then update the key while holding the cache lock.

File: quic-client/src/lib.rs [hunk 11]
```diff
@@ -140,9 +142,14 @@ impl QuicConfig {
     pub fn update_client_certificate(
         &mut self,
         keypair: &Keypair,
-        ipaddr: IpAddr,
+        ipaddr: Option<IpAddr>,
```
Comment (behzadnouri) [inline quic-client/src/lib.rs:145]: The `Option<IpAddr>` is ambiguous where it is fine to have `None` and where it is not.

Instead of overloading this method, can we keep `ipaddr: IpAddr` here, and instead add a separate `update_keypair` method which does not take `ipaddr` argument?

Also doing so, the `update_keypair` method will only change `self.client_certficate`.
If `self.client_certificate` is wrapped in `RwLock` then the method receiver will be an immutable `&self`.
That would remove the need to wrap the `ConnectionManger` in a `RwLock`.

**Review by behzadnouri — COMMENTED** [review]
lgtm, but can you also please ask @lijunwangs to take a look and confirm there is no perf concern?
maybe also run the bench-tps with this change just to be safe.

File: core/src/validator.rs [hunk 12]
```diff
@@ -495,7 +496,7 @@ impl Validator {
         tpu_connection_pool_size: usize,
         tpu_enable_udp: bool,
         admin_rpc_service_post_init: Arc<RwLock<Option<AdminRpcRequestMetadataPostInit>>>,
-    ) -> Result<Self, String> {
+    ) -> Result<(Self, Vec<Arc<dyn NotifyKeyUpdate + Sync + Send>>), String> {
```
[the remark this answers is not in the document]
Comment (ryleung-solana) [reply core/src/validator.rs:499]: It's not just Validator that has to return a key updater. This just seems like the most convenient and simple way to get a list of key updaters to where they're needed. Also if we add a method to get the key updater from Validator, we'd have to store the Vec of updaters somewhere, which is useless, and it's not clear why changing the startup order is a problem.

**Comment by ryleung-solana** [comment]
> lgtm, but can you also please ask @lijunwangs to take a look and confirm there is no perf concern? maybe also run the bench-tps with this change just to be safe.


Ran bench-tps, perf characteristics look pretty much the same. First is with the change, second without.
Screenshot from 2023-12-02 03-36-52


Screenshot from 2023-12-02 05-00-08

File: core/src/validator.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lijunwangs) [reply core/src/validator.rs:499]: There is simpler way to achieve your goal without modifying the Validator::new return interface. Can you try to use AdminRpcRequestMetadataPostInit which is actually used when updating the identity key by the admin RPC service?

File: quic-client/src/lib.rs [hunk 13]
```diff
@@ -143,7 +156,23 @@ impl QuicConfig {
...
+        let mut cert_guard = self.client_certificate.write().unwrap();
+
+        *cert_guard = Arc::new(QuicClientCertificate {
+            certificate: cert,
+            key: priv_key,
+        });
+        Ok(())
+    }
+
+    pub fn update_keypair(&self, keypair: &Keypair) -> Result<(), RcgenError> {
```
Comment (lijunwangs) [inline quic-client/src/lib.rs:170]: This update works only when a new connection pool is created. How do we have handle existing connection cached?

Comment (ryleung-solana) [reply quic-client/src/lib.rs:170]: See connection-cache/src/connection_cache.rs line 141

Comment (lijunwangs) [reply quic-client/src/lib.rs:170]: I understand that, but I think it won't fix the existing connection's endpoints. We would similar things you did to the server side: like 

        let (config, _) = configure_server(key, self.gossip_host)?;
        self.endpoint.set_server_config(Some(config));
        Ok(())

**Comment by ryleung-solana** [comment]
Connection refs are only held by callers to get_connection as long as they
are used to do a send task, afterwards they are dropped. The endpoint is
configured the first time a connection pool is used. Therefore by clearing
the map, the first time get_connection is called afterwards, a new
connection pool is created, and the first time a connection is obtained
from the connection pool the endpoint will be given the new configuration.

**Comment by lijunwangs** [comment]
> Connection refs are only held by callers to get_connection as long as they are used to do a send task, afterwards they are dropped. The endpoint is configured the first time a connection pool is used. Therefore by clearing the map, the first time get_connection is called afterwards, a new connection pool is created, and the first time a connection is obtained from the connection pool the endpoint will be given the new configuration.
> […](#)

Right, I missed the part you cleared the cache.

**Review by lijunwangs — APPROVED** [review]
Looks good to me
